task_manager/task_manager.hh

```diff
@@ -0,0 +1,362 @@
+        }
+
+        void finish() noexcept {
+            if (!_impl->_done.available()) {
+                _impl->_status.end_time = db_clock::now();
+                _impl->_status.state = task_manager::task_state::done;
+                _impl->_done.set_value();
+            }
+        }
+
+        void finish_failed(std::exception_ptr ex, std::string error) noexcept {
+            if (!_impl->_done.available()) {
+                _impl->_status.end_time = db_clock::now();
+                _impl->_status.state = task_manager::task_state::failed;
+                _impl->_status.error = std::move(error);
+                _impl->_done.set_exception(ex);
+            }
+        }
+
+        void finish_failed(std::exception_ptr ex) {
+            finish_failed(ex, fmt::format("{}", ex));
+        }
+    };
+
+    class module : public enable_shared_from_this<module> {
+    protected:
+        task_manager& _tm;
+        std::string _name;
+        task_map _tasks;
+        seastar::gate _gate;
+        uint64_t _sequence_number = 0;
+    public:
+        module(task_manager& tm, std::string name) noexcept : _tm(tm), _name(std::move(name)) {}
+
+        uint64_t new_sequence_number() noexcept {
+            return ++_sequence_number;
+        }
+
+        task_manager& get_task_manager() noexcept {
+            return _tm;
+        }
+
+        virtual abort_source& abort_source() noexcept {
+            return _tm.abort_source();
+        }
+
+        seastar::gate& async_gate() noexcept {
+            return _gate;
+        }
+
+        const std::string& get_name() const noexcept {
+            return _name;
+        }
+
+        task_manager::task_map& get_tasks() noexcept {
+            return _tasks;
+        }
+
+        const task_manager::task_map& get_tasks() const noexcept {
+            return _tasks;
+        }
+
+        void register_task(task_ptr task) {
+            _tasks[task->id()] = task;
+            try {
+                _tm.register_task(task);
+            } catch (...) {
+                _tasks.erase(task->id());
+                throw;
+            }
+        }
+
+        void unregister_task(task_id id) noexcept {
+            _tasks.erase(id);
+            _tm.unregister_task(id);
+        }
+
+        virtual future<> stop() noexcept {
+            tmlogger.info("Stoppping module {}", _name);
+            co_await _gate.close();
+            _tm.unregister_module(_name);
+        }
+
+        template<typename T>
+        requires std::is_base_of_v<task_manager::task::impl, T>
+        future<task_id> make_task(unsigned shard, task_id id = task_id::create_null_id(), std::string keyspace = "", std::string table = "", std::string type = "", std::string entity = "", task_id parent_id = task_id::create_null_id()) {
+            foreign_task_ptr parent;
+            uint64_t sequence_number = 0;
+            if (parent_id) {
+                parent = co_await lookup_task_on_all_shards(_tm.container(), parent_id);
```

Comment:
Can we fold shard it into task id somehow, to avoid this lookup?

Reply:
I created a `parent_data` struct with parent task id and its shard, that is passed here. In most cases the caller would have an access to parent shard anyway so its almost free. Only in test api we still need to use `lookup_task_on_all_shards` not to put a responsibility for adding a parent shard on user. I hope you find that solution ok.

We could also hide the shard id into `task_id`, that would be rather cumbersome though, since we would need to care if the chosen bits aren't used for anything else (least significant bits are not I think), chose the "limit" for shards number, and made sure that we do not disrupt the hash distribution.